nixos/modules/security/pam/default.nix

```diff
@@ -0,0 +1,190 @@
+# This module provides configuration for the PAM (Pluggable
+# Authentication Modules) system.
+
+{ config, lib, pkgs, utils, ... }:
+
+with lib;
+
+let
+  cfg = config.security.pam;
+
+  pamEntryModule = entryType: { config, ... }: {
+    options = {
+      entryType = mkOption {
+        type = types.enum utils.pam.entryTypes;
+        default = entryType;
+        internal = true;
+        description = ''
+          The PAM type for this entry. This should not be set manually.
+        '';
+      };
+
+      control = mkOption {
+        type = utils.pam.controlType;
+        description = ''
+          The PAM control for this entry. It can be either on of the
+          historical basic control keywords, of a set of result-actions pairs.
+        '';
+        apply = value: if isString value then value else
```

Comment:
Don't use `apply`, it leaks the module implementation to the outside, since `config.the.option`, accessible from all modules, will have this function applied already. This also makes it difficult for other modules to reuse this option's value. In this case parsing would be needed.

You can achieve this with a `let in` binding instead.

Same for all the other uses of `apply`